feat(agenttoolkit): Adding thinking steps like in Deep Search

visuniq opened the pull request:
## 🚀 Summary
Refs: (https://unique-ch.atlassian.net/browse/UN-13939)

- Adding a steps log like it was done for Deepsearch
-  Added the logger_messager Class.
- Added this class into orchestrator + unique_websearch + unique_internal_search
- We still use the old logger (ignoring the new fields for now which only will come in a new PI)
- Additionally see the picture and the test instruction.

Briefly describe the changes introduced in this PR.

## ✅ Changes

- Added a new Class called MessageStepLogger to the agent subfolder in the ai repo
- This MessageStepLogger writes and updates progress in the Webservice, The internalSearchTool and the Orchestrator.
- In this version, we created logs for some core processes in the Orchestrator (very limited)
- And produce a post for each tool call, including the returned url and the asked queries/questions
- The strange circle symbol is also coming when the list is empty. --> Therefore, it is not caused by this update. I expect it being shown in cases with not enough returned pdf or webpages.


## 🧪 Testing

Like you see below, we have here a new step log. 
<img width="1414" height="757" alt="Screenshot 2025-11-17 at 10 14 03" src="https://github.com/user-attachments/assets/50f1521b-fdf8-47bd-9b57-6cf64861d782" />

- Start your env with an agent which allows you to at least select WebSearch and InternalToolSearch
- Search something like :  "I really wonder about Meta, Earnings composition, which part dominates"
- Naturally, it is best for this test to have Forced tools for Internal Search, Websearch (alone and both at the same time)
- + Some Internal entries about Meta, for example, you could load up the annual report of meta.
- This should return the result above.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces a MessageStepLogger and integrates message-log entries (with ordered steps and references) into WebSearch, InternalSearch, and the Orchestrator, updating WebSearch executors to return log metadata and adding comprehensive tests.
> 
> - **Toolkit**:
>   - **New `MessageStepLogger`**: Adds per-message ordered logging of steps with details and references (`agentic/logger_manager`), exported and fully tested.
>   - **Tools Base**: Injects `_message_step_logger` into `Tool` for easy logging in tools.
> - **Orchestrator**:
>   - Logs key phases: start, each loop iteration, and answer generation; wires `MessageStepLogger` via builder.
> - **WebSearch**:
>   - **Executors API**: `run()` now returns `(content_chunks, queries_for_log)`; adds `WebSearchLogEntry` and reference conversion in `WebSearchResult.to_content_reference`.
>   - **Service**: Creates message log entries labeled "Web Search" with `MessageLogDetails` and compiled `ContentReference`s.
>   - **Tests**: Add/adjust tests for descriptions, executors selection, run flow, progress notifications, and message-log preparation.
> - **InternalSearch**:
>   - **Service/Tool**: Logs "Internal Search" entries with queries and internal `ContentReference`s; debug key changed to `searchStrings` (plural).
>   - **Tests**: New tests validate reference list creation, sequence/order, naming, and URL handling.
> - **Versions/Changelogs**: Bump `unique_toolkit` to `1.27.2`, `unique_orchestrator` to `1.7.12`, `unique_web_search` to `1.6.0`, and `unique_internal_search` to `1.2.5`, noting message log inclusion.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5901fac13ec9a96d16da557d2f22e7e1f771da13. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

visuniq committed:
Task: UN-13939

Target:
- Adding a steps log like Gustav did it for Deepsearch
- I got the command to add a new class which is later loging even more steps.

This class is only a predraft since we will replace big parts of the underlying message.logger with the help of Richards team in one of the next PI

- Tested with automatic tests
- Tested in local env and resulting in something at least good.

visuniq committed:
Task: UN-13939

Target:
- Adding a steps log like Gustav did it for Deepsearch
- I got the command to add a new class which is later loging even more steps.

This class is only a predraft since we will replace big parts of the underlying message.logger with the help of Richards team in one of the next PI

- Tested with automatic tests
- Tested in local env and resulting in something at least good.

klcd commented on unique_orchestrator/unique_orchestrator/unique_ai.py:
```diff
@@ -236,6 +250,9 @@ async def _process_plan(self, loop_response: LanguageModelStreamResponse) -> boo
         self._logger.debug("No tool calls. we might exit the loop")
 
+        ## No tool calls means we wrap it up, at least in my understanding.
+        self.messagesteplogger.write_message_log_text_message("**Final result aggregation**")
```
keywordargument

klcd commented on tool_packages/unique_web_search/unique_web_search/service.py:
```diff
@@ -97,11 +99,14 @@ def evaluation_check_list(self) -> list[EvaluationMetricName]:
             tool_call.arguments,
         )
 
+        #print(parameters)
```
This should not land in production code

klcd commented on unique_toolkit/unique_toolkit/agentic/logger_manager/service.py:
```diff
@@ -0,0 +1,205 @@
+        self._event: ChatEvent = event
+
+    @staticmethod
+    def create_message_log_entry(
```
Why is this necessary? Only used within the class, I would make this an internal classmethod

visuniq replied:
This seems to work. Hmm, I guess the reason was that the coding grew "organic" as part of my onboarding.

klcd replied:
I do not understand what you mean by this seems to work. Let me know tomorrow

visuniq committed:
Moving the test file to it's correct place.

- tests moving around
- Changelog

visuniq committed:
Moving the test file to it's correct place.

- tests moving around
- Changelog

visuniq commented on unique_orchestrator/unique_orchestrator/unique_ai.py:
```diff
@@ -96,13 +100,21 @@ async def run(self):
         ## Loop iteration
         for i in range(self._config.agent.max_loop_iterations):
+            
+            ii = i + 1
```
It counts otherwise from 0 which looks super bad in GUI

visuniq commented on unique_orchestrator/unique_orchestrator/unique_ai.py:
```diff
@@ -96,13 +100,21 @@ async def run(self):
             )
 
         ## Loop iteration
         for i in range(self._config.agent.max_loop_iterations):
```
This may be but it is rather part of our general refactoring? and not of thes task?

visuniq committed:
graphical adaptions like wished by Fabian

Adaptions, better and clearer log

visuniq committed:
graphical adaptions like wished by Fabian

Adaptions, better and clearer log

visuniq commented on unique_orchestrator/unique_orchestrator/unique_ai.py:
```diff
@@ -82,6 +85,7 @@ def __init__(
         self._mcp_servers = mcp_servers
         self._streaming_handler = streaming_handler
 
+        self.messagesteplogger = messagesteplogger
```
Done

visuniq committed:
pyproject toml

visuniq committed:
pyproject toml

visuniq committed:
bidantic fixes part 1

visuniq committed:
bidantic fixes part 1

visuniq committed:
tests fixed for BasedPyright rules (lint)

visuniq committed:
tests fixed for BasedPyright rules (lint)

visuniq committed:
tried to fix git ruff by locally running ruff

visuniq committed:
tried to fix git ruff by locally running ruff

visuniq committed:
ruff --fix

visuniq committed:
ruff --fix

visuniq committed:
changes by Ruff which seem to not affect tests and real work in any way.

visuniq committed:
changes by Ruff which seem to not affect tests and real work in any way.

visuniq committed:
last ruff file change

visuniq committed:
last ruff file change

unique-fabian commented on unique_toolkit/CHANGELOG.md:
```diff
@@ -5,6 +5,10 @@ All notable changes to this project will be documented in this file.
+## [1.25.2222] - 2025-11-13
+- MessageLogger Class
+
+
```
remove line

visuniq committed:
minor visual changes

- Dot points for questions
- No additional comment for questions by AI.
- Iteration instead of iteration

visuniq committed:
minor visual changes

- Dot points for questions
- No additional comment for questions by AI.
- Iteration instead of iteration

unique-fabian commented on unique_toolkit/unique_toolkit/agentic/logger_manager/service.py:
```diff
@@ -0,0 +1,238 @@
+
+    This class provides utilities for creating message log entries with proper
+    ordering, references, and details. It extends the functionality from
+    deepsearch utils to be usable across all agentic tools.
```
also here remove link to deepsearch

unique-fabian commented on unique_toolkit/unique_toolkit/agentic/logger_manager/service.py:
```diff
@@ -0,0 +1,238 @@
+        """
+
+        count = 0
+        data: list[ContentReference] = []
```
input parameter data is overwritten

unique-fabian commented on unique_toolkit/unique_toolkit/agentic/logger_manager/service.py:
```diff
@@ -0,0 +1,238 @@
+                reference_name = content_chunk.title
+            else:
+                reference_name = content_chunk.key or ""
+
```
can be simplified to a one-line

`reference_name = content_chunk.title or content_chunk.key or ""`

visuniq committed:
hard enforced keyword parameters

- Not only proper keyword parameters in my calling of functions but
- Enforcement of keyword parameters.

visuniq committed:
hard enforced keyword parameters

- Not only proper keyword parameters in my calling of functions but
- Enforcement of keyword parameters.

unique-fabian commented on unique_orchestrator/unique_orchestrator/unique_ai.py:
```diff
@@ -136,6 +147,8 @@ async def run(self):
                 self.start_text, loop_response
             )
 
+        ## Placeholder to set end of process label, if wished.
```
remove this comment if we are not doing anything

unique-fabian commented on tool_packages/unique_internal_search/unique_internal_search/service.py:
```diff
@@ -164,6 +174,14 @@ async def search(
                 )
             )
 
+        # Updating our logger with the search results for all search strings.
```
update or remove description. we are not just displaying the search strings, but also the reuslts

unique-fabian commented on tool_packages/unique_internal_search/unique_internal_search/service.py:
```diff
@@ -164,6 +174,14 @@ async def search(
 
+        # Updating our logger with the search results for all search strings.
+        if self.message_step_logger:
+            self.message_step_logger.create_full_specific_message(
```
my proposal would be to include the define_reference_list_for_internal function into the create_full_specific_message. like this we give all parameters as input and then the data generation is all handled inside

unique-fabian commented on tool_packages/unique_internal_search/unique_internal_search/service.py:
```diff
@@ -283,6 +302,10 @@ def __init__(
 
+        # Initialize MessageStepLogger if event is a ChatEvent
+        # if isinstance(self.event, (ChatEvent, Event)):
+        self.message_step_logger = MessageStepLogger(self._chat_service, self._event)
```
why you initialize at the end? not needed, correct?

visuniq replied:
It is, it gives no error, but it not shows up in testing if I not do this. It confused me too. Maybe related with the thing I should discuss now with Rami.

unique-fabian commented on tool_packages/unique_web_search/unique_web_search/service.py:
```diff
@@ -110,6 +114,21 @@ async def run(self, tool_call: LanguageModelFunction) -> ToolCallResponse:
+            # Write entry with found hits, WebSearch V1 has just one call.
+            data: list[ContentReference] = (
+                self.message_step_logger.define_reference_list(
+                    source="web", content_chunks=content_chunks, data=[]
```
same comments as for internal search

visuniq committed:
Fixes as requested by Fabian in Review.

visuniq committed:
Fixes as requested by Fabian in Review.

visuniq committed:
remove no longer logically needed field

visuniq committed:
remove no longer logically needed field

visuniq committed:
updated tests for last minor change with no needed source variable.

visuniq committed:
updated tests for last minor change with no needed source variable.

unique-fabian commented on tool_packages/unique_internal_search/unique_internal_search/service.py:
```diff
@@ -84,6 +81,50 @@ async def is_chat_only(self, **kwargs) -> bool:
+        """
+        count = 0
+        for content_chunk in content_chunks:
+            reference_name: str
```
as mentioned in previous review. this can be a one liner

`reference_name = content_chunk.title or content_chunk.key or ""`

unique-fabian commented on tool_packages/unique_web_search/unique_web_search/service.py:
```diff
@@ -97,6 +95,48 @@ def tool_format_information_for_system_prompt(self) -> str:
+        """
+
+        count = 0
+        data: list[ContentReference] = []
```
remove

visuniq committed:
cleaning up changes.

visuniq committed:
cleaning up changes.

visuniq committed:
True general message logger class.

visuniq committed:
True general message logger class.

visuniq committed:
format minor fix

visuniq committed:
format minor fix

visuniq committed:
last *

visuniq committed:
last *

visuniq committed:
lint

visuniq committed:
lint

visuniq committed:
linter bug

visuniq committed:
linter bug

visuniq commented on tool_packages/unique_internal_search/unique_internal_search/service.py:
```diff
@@ -164,6 +219,18 @@ async def search(
+                data=[MessageLogEvent(type="InternalSearch", text="")]
+            ),
+            data=data,
+        )
```
Yes, this is an open question for me. This was the reason for why I had initialization at the what you called bottom first before setting it within the tool class. It is a complex question related to a somehow different build of Internal Search versus Web Search or even Deep tools. I am honestly not sure about it.

visuniq commented on tool_packages/unique_internal_search/unique_internal_search/service.py:
```diff
@@ -81,6 +85,48 @@ async def is_chat_only(self, **kwargs) -> bool:
+                    )
+                )
+
+            count += 1
```
done

visuniq committed:
Cursor bot comments

visuniq committed:
Cursor bot comments

visuniq commented on tool_packages/unique_web_search/unique_web_search/service.py:
```diff
@@ -94,9 +99,52 @@ def tool_format_information_for_system_prompt(self) -> str:
+                )
+                count += 1
+
+        return data
```
done

visuniq committed:
last minor fix

visuniq committed:
last minor fix

visuniq commented on unique_toolkit/unique_toolkit/agentic/logger_manager/service.py:
```diff
@@ -0,0 +1,87 @@
+            _request_counters[message_id] = 0
+
+        _request_counters[message_id] += 1
+        return _request_counters[message_id]
```
Like mentioned, this method exists in deepsearch but it is only called in tests. This means, it has no effect.

unique-fabian commented on tool_packages/unique_web_search/unique_web_search/service.py:
```diff
@@ -110,6 +158,25 @@ async def run(self, tool_call: LanguageModelFunction) -> ToolCallResponse:
+                content_chunks=content_chunks,
+                data=[],
+            )
+            # Only log for V1 mode (WebSearchToolParameters has query, WebSearchPlan doesn't)
```
Did you talk to @ThePhilAz ? We also need a solution for V2 mode as this will be very soon the default

unique-fabian committed:
inputs Fabian

unique-fabian committed:
amend

unique-fabian committed:
amend

visuniq committed:
Merge branch 'viri_notifier_search' of https://github.com/Unique-AG/ai into viri_notifier_search

unique-fabian committed:
move message log out from service to tool. (internal search)

unique-fabian committed:
Merge branch 'viri_notifier_search' of github.com:Unique-AG/ai into viri_notifier_search

raheemazeezabiodun commented on tool_packages/unique_web_search/unique_web_search/service.py:
```diff
@@ -94,6 +99,47 @@ def tool_format_information_for_system_prompt(self) -> str:
+        message = ""
+        for entry in query_list:
+            message += f"• {entry}\n"
+        message = message.strip("\n")
```
```suggestion
        message = "\n".join(f"• {entry}" for entry in query_list)
```
same as above

raheemazeezabiodun reviewed (commented):
looks good in general, left some comments

visuniq committed:
Internal Search using the question logger subfield + tests + formating

- Removed my logger methods from WebSearch since this will be partially replaced with alternative code.

visuniq committed:
Web Search with Websearch V2

- After long discussion with Rami, we have a new concept of getting the values.

visuniq committed:
fixes before tests and ruff

visuniq committed:
Tests, needed str enforcement and ruff formatting

visuniq committed:
Type error problem

This would fix the bug report my logs with type error or WebSearchResult, it looks despite this correct to just send the string.

visuniq commented:
No longer relevant and merged after copy and rebasing months ago.
